src/Tuura/Plato/Translation.hs

```diff
@@ -19,36 +25,50 @@ instance Ord Signal
     where
         compare (Signal x) (Signal y) = compare x y
 
-addErrors :: (Eq a, Show a) => [a] -> [a] -> [a] -> [a] -> String
-addErrors unused incons undef invInit = un ++ ic ++ ud ++ iv
-  where
-    un = if (unused  /= []) then ("The following signals are not declared as input, output or internal: \n" ++ unlines (map show unused) ++ "\n") else ""
-    ic = if (unused  /= []) then ("The following signals have inconsistent inital states: \n" ++ unlines (map show incons) ++ "\n") else ""
-    ud = if (undef   /= []) then ("The following signals have undefined initial states: \n" ++ unlines (map show undef) ++ "\n") else ""
-    iv = if (invInit /= []) then ("The initial state is an invariant state. Signals with erroneous initial states: \n" ++ unlines (map show invInit) ++ "\n") else ""
+addErrors :: Show a => ValidationError a -> String
+addErrors err = case err of
+        UnusedSignal as ->
+            "The following signals are not declared as input, output or internal: \n" ++ unlines (signs as) ++ "\n"
+        InconsistentInitialState as ->
+            "The following signals have inconsistent inital states: \n" ++ unlines (signs as) ++ "\n"
+        UndefinedInitialState as ->
+            "The following signals have undefined initial states: \n" ++ unlines (signs as) ++ "\n"
+        InvariantInitialState as ->
+            "The initial state is an invariant state.\nSignals with erroneous initial states: \n" ++ unlines (signs as) ++ "\n"
+    where
+        signs = map show
+
+  --   un ++ ic ++ ud ++ iv
+  -- where
+  --   un = if (unused  /= []) then ("The following signals are not declared as input, output or internal: \n" ++ unlines (map show unused) ++ "\n") else ""
+  --   ic = if (unused  /= []) then ("The following signals have inconsistent inital states: \n" ++ unlines (map show incons) ++ "\n") else ""
+  --   ud = if (undef   /= []) then ("The following signals have undefined initial states: \n" ++ unlines (map show undef) ++ "\n") else ""
+  --   iv = if (invInit /= []) then ("The initial state is an invariant state.\nSignals with erroneous initial states: \n" ++ unlines (map show invInit) ++ "\n") else ""
 
 validate :: Ord a => [a] -> CircuitConcept a -> ValidationResult a
 validate signs circuit
     | unused ++ inconsistent ++ undef ++ invInit == [] = Valid
-    | otherwise                             = Invalid unused inconsistent undef invInit
+    | otherwise = Invalid (
+        (if unused       /= [] then [UnusedSignal unused]                   else []) ++
```

Comment:
With the above comment, you would be able to write `map UnusedSignal unused` instead of this line.